### PR TITLE
Avoid integer overflow when parsing cbor negative numbers

### LINKED ISSUE #124: [cbor] Invalid value returned for negative int32 where the absolute value is > 2^31 - 1
In https://github.com/FasterXML/jackson-dataformats-binary/issues/30, parsing of positive integers was fixed for the case where the uint32 value overflows a java int: https://github.com/FasterXML/jackson-dataformats-binary/blob/d8f82c788c4d8c41c59ca143caf2f989d0e9a9ae/cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java#L679-L689

However, the parsing of negative numbers (major type 1) was not adjusted correctly:

https://github.com/FasterXML/jackson-dataformats-binary/blob/d8f82c788c4d8c41c59ca143caf2f989d0e9a9ae/cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java#L722-L731

The cast of `_numberLong = ((long) v) + -1L;` does not adequately prevent overflow; if v is negative, it has to be adjusted to represent the uint32 value, and then negated:

```
case 2:
    {
        int v = _decode32Bits();
        if (v < 0) {
            long unsignedV = (long) v & 0xFFFFFFFFL;
            _numberLong = -1L - unsignedV;
            _numTypesValid = NR_LONG;
        } else {
            _numberInt = -v - 1;
        }
    }
    break;
```

Similar adjustment is required for other sizes.

As a test case, deserializing `3A 9FF0947F` should produce `-2683344000`; currently it results in `-1611623298`.

### PR DESCRIPTION
**Pull request by JacekLach**
Fixes #124 

I couldnt get the master branch to build:

```
[FATAL] Non-resolvable parent POM for com.fasterxml.jackson.dataformat:jackson-dataformats-binary:3.0.0-SNAPSHOT: Could not find artifact com.fasterxml.jackson:jackson-base:pom:3.0.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 3, column 11
```

So I made this against 2.9. I expect it should be easy to move to master if desired.

### CODE REVIEW

**Comment by cowtowncoder** [comment]
To build master you need to build some of the dependencies locally. I can try to see if there's a way to push snapshot builds of some of the deps.
But PR for 2.9 makes sense regardless since I can merge forward easily (but not back from master).

Thank you for submitting this.

File: cbor/src/test/java/com/fasterxml/jackson/dataformat/cbor/parse/ParserNumbersTest.java [hunk 1]
```diff
@@ -170,6 +181,21 @@ public void testInt64Overflow() throws Exception
         assertEquals(exp, p.getBigIntegerValue());
         assertEquals(NumberType.BIG_INTEGER, p.getNumberType());
         p.close();
+
+        // and, combined, a negative number where the mantissa overflows a signed int32
+        input = new byte[] {
+                (byte) CBORConstants.PREFIX_TYPE_INT_NEG + 27, // int32, that is, 4 more bytes
```
Comment (cowtowncoder) [inline cbor/src/test/java/com/fasterxml/jackson/dataformat/cbor/parse/ParserNumbersTest.java]: Wouldn't this be `int64`?

**Comment by cowtowncoder** [comment]
Ok, I will have to read this through a few times to fully grok it. I assume tests make sense but it seems like there should be slightly simpler coercions for edge cases (avoid one operation at least).

File: cbor/src/test/java/com/fasterxml/jackson/dataformat/cbor/parse/ParserNumbersTest.java (hunk 1 above)
Comment (JacekLach) [reply cbor/src/test/java/com/fasterxml/jackson/dataformat/cbor/parse/ParserNumbersTest.java]: You're right, bad copy paste. I fixed the constant but not the comment. will update later

**Comment by JacekLach** [comment]
Possibly; I just did the same thing that was done for the positive numbers.
